awx/main/models/unified_jobs.py

```diff
@@ -973,22 +977,42 @@ def create_config_from_prompts(self, kwargs, parent=None):
             valid_fields.extend(['survey_passwords', 'extra_vars'])
         else:
             kwargs.pop('survey_passwords', None)
+        many_to_many_fields = []
         for field_name, value in kwargs.items():
             if field_name not in valid_fields:
                 raise Exception('Unrecognized launch config field {}.'.format(field_name))
-            if field_name == 'credentials':
+            if isinstance(getattr(self.__class__, field_name).field, models.ManyToManyField):
+                many_to_many_fields.append(field_name)
+                continue
+            if isinstance(getattr(self.__class__, field_name).field, (models.ForeignKey)):
+                if value:
+                    setattr(config, "{}_id".format(field_name), value.id)
                 continue
             key = field_name
             if key == 'extra_vars':
                 key = 'extra_data'
             setattr(config, key, value)
         config.save()
 
-        job_creds = set(kwargs.get('credentials', []))
-        if 'credentials' in [field.name for field in parent._meta.get_fields()]:
-            job_creds = job_creds - set(parent.credentials.all())
-        if job_creds:
-            config.credentials.add(*job_creds)
+        for field_name in many_to_many_fields:
+            if field_name == 'credentials':
+                # Credentials are a special case of many to many because of how they function
+                # (i.e. you can't have > 1 machine cred)
+                job_item = set(kwargs.get(field_name, []))
+                if field_name in [field.name for field in parent._meta.get_fields()]:
+                    job_item = job_item - set(getattr(parent, field_name).all())
+                if job_item:
+                    getattr(config, field_name).add(*job_item)
+            else:
+                # Here we are doing a loop to make sure we preserve order in case this is a Ordered field
+                job_item = kwargs.get(field_name, [])
+                if job_item:
+                    parent_items = list(getattr(parent, field_name, []).all())
+                    for item in job_item:
+                        # Do not include this item in the config if its in the parent
+                        if item not in parent_items:
+                            getattr(config, field_name).add(item)
```

Comment:
The "Ordered field" block here can't be right - it contradicts the prompting rules we wrote for `instance_groups`.

If the launch-time list is not `[]`, then _the entire list_ constitutes the prompted value. This loop obtains any items _added relative to the JT_ and that's completely irrelevant of what it actually does. _Even if the entire list matches the JT_, if you provide it on launch, that gives a different meaning. The JT `instance_groups` says "use these groups, but fallback to inventory/org". The prompted list says "_only_ use these groups".

Putting it that way, it's also clear that the `default` value offered in the `JobLaunchSerializer` is wrong, because it returns the JT instance groups. Using that as-is would use that same list, but with a different meaning to it. So either the entire construct of the `default` should be declared invalid for this field, or the view should present `"default": []` or `"default": null`, because this is what obtains parity with the basic JT behavior.


While the prompts store is only used for the obscure purposes of relaunch or the "create_schedule", this makes obvious another more likely and severe bug. The method `create_unified_job` takes `kwargs` with an unclear meaning. Even if the UI gives `[]`, that doesn't mean we're off the hook because `copy_m2m_relationships` takes _any_ value in kwargs to be an _override_, including a falsy value, so it's not even clear that our our intended not-provided value will leave the JT values untouched.

I'm sure testing is at high risk of missing these things, because we are very light on testing (if at all) filling in JT values and then assuring launch works _without_ providing prompts, when prompts are accepted. Even if we did, integration tests use its own client, and it's not clear even what payload to the launch POST they are even _supposed_ to provide for this case, because in the request data, you can _exclude_ the key, set to _null_, set to `[]`, or set to the "default" values. I don't think we have ever touched that testing matrix.

Reply:
I don't want to leave this too vague...

I'm saying that, as you have already added special logic in `copy_m2m_relationships`, you need more special logic. Before you start the loop `for ig_id in override_field_val:` in the special IG logic, it needs to be checked that the data constitute an override in the first place.... like, indent that into a `if override_field_val:` conditional. But I'd hate to jump right into that without first characterizing relevant test cases.

Reply:
Oh, and this is actually impossible / impractical to test in integration tests. Only unit tests, because the resolved list is never reported by the API.

Reply:
Scratch some of that, don't take my prior advice. In the event that IGs are not provided, IIRC, we do not want to copy the JT values at all. Again, this is philosophically very problematic with `copy_m2m_relationships`, and it should probably be done before that, but using the existing code structure, that method should _always_ `continue` on the instance_groups field to avoid a copy of the JT list. I do need to articulate a specific test case for this.

Reply:
We discussed this and I am developing a potential patch.

https://github.com/ansible/awx/pull/12925

Needs work post-migration merge.